### ISSUE TITLE
Decorators for special component API

### ISSUE THREAD

**Issue by buschtoens** [ember-decorators/ember-decorators#133]
https://github.com/rwjblue/ember-decorators/issues/123#issuecomment-316494043:
> We should add some decorators specifically for handling those special component API's (tagName, classNames, classNameBindings, are there others?). Can you open a separate issue for that?

- [x] [`attributeBindings`](https://www.emberjs.com/api/ember/2.14/classes/Ember.Component/properties/attributeBindings?anchor=attributeBindings) from [`ViewMixin`](https://www.emberjs.com/api/ember/2.14/classes/Ember.ViewMixin)
- [x] [`classNameBindings`](https://www.emberjs.com/api/ember/2.14/classes/Ember.Component/properties/classNameBindings?anchor=classNameBindings) from [`ClassNamesSupport`](https://www.emberjs.com/api/ember/2.14/classes/Ember.ClassNamesSupport)
- [x] [`classNames`](https://www.emberjs.com/api/ember/2.14/classes/Ember.Component/properties/classNames?anchor=classNames) from [`ClassNamesSupport`](https://www.emberjs.com/api/ember/2.14/classes/Ember.ClassNamesSupport)
- [ ] [`elementId`](https://www.emberjs.com/api/ember/2.14/classes/Ember.Component/properties/elementId?anchor=elementId) from [`ViewMixin`](https://www.emberjs.com/api/ember/2.14/classes/Ember.ViewMixin)
- [ ] [`layout`](https://www.emberjs.com/api/ember/2.14/classes/Ember.Component/properties/layout?anchor=layout)
- [x] [`tagName`](https://www.emberjs.com/api/ember/2.14/classes/Ember.Component/properties/tagName?anchor=tagName) from [`ViewMixin`](https://www.emberjs.com/api/ember/2.14/classes/Ember.ViewMixin)

**Comment by buschtoens** [comment]
@rwjblue How do you imagine turning these into decorators?

The general problem I see, is that there are potentially more [`concatenatedProperties`](https://www.emberjs.com/api/ember/2.14/classes/Ember.Object/properties/concatenatedProperties?anchor=concatenatedProperties) and [`mergedProperties`](https://www.emberjs.com/api/ember/2.14/classes/Ember.Object/properties/mergedProperties?anchor=mergedProperties) that we do not / cannot anticipate. For instance, @dfreeman's brilliant [`ember-css-modules`](https://github.com/salsify/ember-css-modules) adds `localClassNames` and `localClassNameBindings` to that list.

**Comment by dfreeman** [comment]
@buschtoens Decorators are actually a pretty great fit for this problem — I imagine ember-css-modules would just expose `localClassNames` and `localClassNameBindings` decorators. 

The cool thing about this is it means we wouldn't have to monkey patch `Ember.Component` like we're doing today, since the class resolution logic could live in/be installed by the decorators themselves.

**Comment by pzuraq** [comment]
I think we can actually generalize this into a `merged` and a `concat` decorator which indicate that the property should opt into merging with the parent class's property. This make merging opt in and less magical overall

**Comment by pzuraq** [comment]
tagName and layout are a bit different. I need to double check, but I think those might actually work on latest Ember? If not, the issue is they need to be set on the prototype of the class, and class properties are set on the instance.

**Comment by buschtoens** [comment]
For attributes and (local) class names this makes for some pretty nice syntax indeed! :) 

```js
import { Component, attribute, className } from 'ember-decorators/component';
import { localClassName } from 'ember-css-modules';

export default class extends Component {
  @attribute title = 'helpful tooltip';
  @attribute disabled = false;
  @attribute('data-language') language = 'en';

  @className isUrgent = true; // => 'is-urgent'
  @className('active', 'inactive') isActive = false; // => 'inactive'
  @className priority = 'highestPriority'; // => 'highestPriority'
  @className otherClassesArray = ['some-class', 'another-one']; // => 'some-class another-one'
  @className otherClassesString = 'some-class another-one'; // => 'some-class another-one'

  @localClassName isUrgent = true; // => styles['is-urgent']
  @localClassName('active', 'inactive') isActive = false; // => styles['inactive']
  @localClassName priority = 'highestPriority'; // => styles['highestPriority']
  @localClassName otherClassesArray = ['some-class', 'another-one']; // => styles['some-class'], styles['another-one']
  @localClassName otherClassesString = 'some-class another-one'; // => styles['some-class'], styles['another-one']
}
```

The only nitpick I have, is that for unbound class names, you still have to hog a class property.

**Comment by buschtoens** [comment]
And +1 for @pzuraq: IMHO `elementId`, `tagName` and `layout` should work as regular class properties.

**Comment by pzuraq** [comment]
Ah, that's interesting. I was seeing it working more like it does now:

```
export default class extends Component {
  @merged classNames = ['is-urgent']
  @merged classNameBindings = ['isActive:active:inactive', 'priority:highestPriority']
}
```

This would be pretty general and work with all merged properties, but I do like the idea of basically eliminating the need for these properties altogether. I could get behind it 😃

**Comment by buschtoens** [comment]
Actually, I think both would be great.

Even if we don't end up with don't end up with `@attribute` and `@className`, the `@merged` decorator would be a super useful addition for other objects.

Also, Ember currently makes a distinction between [`mergedProperties`](https://www.emberjs.com/api/ember/2.14/classes/Ember.Object/properties/mergedProperties?anchor=mergedProperties=) for Objects and [`concatenatedProperties`](https://www.emberjs.com/api/ember/2.14/classes/Ember.Object/properties/concatenatedProperties?anchor=concatenatedProperties) for Arrays. Should we keep that distinction? In that case you'd have to swap `@merged` with `@concatenated` in your snippet. :wink:

**Comment by dfreeman** [comment]
There's an elegance to `@merged`/`@concatenated`, but I worry that it would be too easy for folks to accidentally forget, or to be fine without it at first and then have things break later because a superclass adds its own `attributeBinding` that you're now overriding.

**Comment by pzuraq** [comment]
@buschtoens yeah, that's where I was originally going with it, but I could also see just having one that detects whether it's an object or array automatically (insert bikeshed here).

@dfreeman I'd actually argue that's a feature, not a bug. One of the main things I've come across with early Ember learners is they are totally confused by the magical behavior of attributeBindings/classNames, etc. Automatically merged properties are convenient sometimes, but super magical. Also, there is no opt out as is, so you're stuck with whatever the superclass decided you should have. Basically, in order to write a subclass effectively, you need to know the entire class hierarchy.

By making the `merged/concatenated` decorators required, this would become opt-in instead and would be explicit. No magic, just a decorator that reads the superclass and modifies the property, plain old Javascript.

**Comment by buschtoens** [comment]
`@merged` / `@concatenated` would have inverted semantics, in that you explicitly have to state that you want to merge / concat that property with the base class's property. I really like that. Especially since it allows you to overwrite that property, which is currently impossible.

To make the transition easier, we (sh|c)ould still honor `concatenatedProperties` and `mergedProperties`, if they were defined on a base class. Or we could count using an ES6 class as an opt-in to ignoring these.

**Comment by pzuraq** [comment]
I'd say ES6 classes are opt-in, especially merging/concat-ing will require changes to CoreObject, so almost definitely has to go through the RFC process. We'll see what comes out of that, but I'm going to push for explicit merging/concatenating.

**Comment by dfreeman** [comment]
> One of the main things I've come across with early Ember learners is they are totally confused by the magical behavior of attributeBindings/classNames, etc. 

I think a key reason for that today is that there's zero indicator that there's _anything_ special about `attributeBindings`, `classNames` and friends. Whether you go with generic `@concatenated` or domain-specific `@className`, either way there's a lexical indicator that this isn't just a normal property.

**Comment by pzuraq** [comment]
Completely agree, the main issue is lack of context. I think both the specific and general decorators both do that, and would be ok with going either direction (or both) in `ember-decorators` - and of course, if we choose one, the other can be done in a separate project since it is just a plain JS decorator. 

My point is that I think this is an opportunity to get rid of some of the magical behavior that makes learning Ember harder, so we shouldn't work to bring back the old behavior in core.

**Comment by dfreeman** [comment]
👍  Definitely not opposed to taking the opportunity to reflect on what's worked and what hasn't, and your point about not being able to opt _out_ of things defined on a superclass is an important one.

It feels like with either approach you wind up needing to know the full class hierarchy though — if you go with `@concatenated`, you need to be aware of whether your ancestors are are applying any classes to decide whether you want to apply the decorator or not.

Taking a step back, components in a Glimmer app don't have an implicit root element, so the need for `attributeBindings`, `classNames`, etc. goes away. There's definitely a camp of people advocating for that to be the behavior for 'glimmer components' in Ember as well, which would ultimately make `layout` the only thing on the original list up top that would still be relevant for new components.

**Comment by rwjblue** [comment]
I think we can do both approaches (in the general sense) in ember-decorators (having an opt-in `merged` / `concat` and having custom `attribute` / `classNames` / etc).

> Taking a step back, components in a Glimmer app don't have an implicit root element, so the need for attributeBindings, classNames, etc. goes away.

Agreed. This is primarily the logic that leads me (personally) towards `attribute` and `classNames` specific decorators, which are then just slowly phased out as glimmer components become more and more common.  

Again, I think its totally fine to add an opt-in mechanism for concat or merged properties, but I see that as a separate issue (generally speaking) from "how do we use these component APIs".

**Comment by buschtoens** [comment]
Okay, if the semantics suggested in https://github.com/rwjblue/ember-decorators/issues/133#issuecomment-316748160 are uncontroversial, the only question that's left is whether to call it `@className` or `@classNames`.

For symmetry with `@attribute` and `classNameBindings`, I would vote for `@className`, since it effectively replaces the latter one. It also offers the same functionality as `classNames` (as a side-effect), but I think the primary use case will be binding class names.

**Comment by dfreeman** [comment]
@buschtoens what would binding a class name to a computed property look like in that proposal?

**Comment by pzuraq** [comment]
I'd imagine you'd just add it to the decorator list

```
@className
@computed
foo() {

}
```

The decorator should just push the name of the property into the `classNameBindings` prop, so it should be able to do it independent of order (just needs the key, won't mutate the descriptor)

**Comment by buschtoens** [comment]
Closed by #166.
